### DISCUSSION TITLE
FFmpeg and deinterlace

### DISCUSSION THREAD

**Discussion by vadegdadeg** [Motion-Project/motion#1606] [Q&A] [answered]
Hello. Dear Mr Dave please tell me how to enable deinterlace in FFmpeg?
I know this option has been removed, but I still have streaks in the frames on my records. I am using analog cameras via SAA7134.
Codec: mkv
Bitrate: 800000
Thank you!

**Comment by Mr-Dave** [comment]
Short version:  movie_extpipe 
Suggestion: netcam_url

Long Version;

I'll be curious as to your results but as I understand it, the issue is likely with the capture rather than the encoding.  Motion captures, processes and then re-encodes frames.  The de-interlace option that previously existed only related to the encoding of movies.  If that is the ONLY spot that the streaks are observed (not via the web interface, picture_output or getting false positives), then the movie_extpipe would allow you to add any encoding options desired.  

But.... I think that the real source of the problem is the capture.  Motion captures frames and decodes them and treats what it gets as "everything" or "complete".  It then compares that captured "full" image to a reference picture and if there are differences, it calls it motion, encodes that single image to a jpg or sends it off to be encoded into a movie.  Whether the original image came from a interlace frame is not even a "thing" anymore when it gets sent to be encoded.  

Therefore, if the camera source is interlaced, every other frame (or whatever frequency exists) captured from the device is a partial frame.  Since this is from the capture device, I would expect everything to be affected including the picture_output, motion detection as well as the web interface.  

So I think the real question should be "how do I get my camera source to be de-interlaced as it is sent into Motion".  For this, I'd suggest using the netcam_url and v4l2 variant.  When the regular video_device option is used, Motion simply pulls the bytes from the device buffer.  If the netcam_url with the v4l2 variant is used, then that gets processed via FFmpeg and with the new netcam_params option, any FFmpeg option desired can be added. 

Finally, I'll caveat the above with.......I do not have one of these capture cards any longer.  This is just my understanding of the interlace process on devices and I could be entirely off base.

**Reply by vadegdadeg** [reply]
Hello!
I apologize for not answering for a long time. I tested different analog capture boards, based on BT878, SAA7134 and EM2860. All of them give an image with a "ladder" on moving objects. I also tried load the m2m-deinterlace kernel module, but this did not help. David, you suggest using a "pipe", as I understand it using v4l2loopback? But this heavily loads the processor, and in the case of three cameras (like mine), the load can be is very high! Can you share with me a "dirty hack" where can enter the magic word 'yadif' in the ffmpeg encoder options in the sources? This is the only rational solution, in my opinion. I am a weak programmer and mainly write code for AVR MCU, in the end I will find where to write this (yadif), but with your help it will be easier to do... Thank you in advance! Sincerely!

**Comment by vadegdadeg** [comment, marked as answer]
I turned out to be a donkey, I was inattentive, forgive me...
I carefully re-read the documentation and realized that movie_extpipe is just the solution to my problem.
After studying ffmpeg a little, I came up with this config:
**First, turn off movie_output**
movie_output off
movie_output_motion off
#movie_quality 0
#movie_bps 800000
#movie_codec mkv
**Then turn on**
movie_extpipe_use on
movie_extpipe ffmpeg -v error -stats -y -f rawvideo -pix_fmt yuv420p -video_size %wx%h -framerate %fps -i pipe:0 -c:v libx265 -crf 21 -x265-params log-level=error -preset ultrafast -vf yadif=2 -f matroska %f.mkv

Where:
-vf yadif=2, filter and filter type, I had timestamp distortion with yadif=1, so yadif=2, but you can choose the value that suits you.
-c:v codec, I decided to try x265, selecting the quality I needed -crf 21 and disabled the output -x265-params log-level=error.
-f matroska, a container that allows viewing unwritten complete files
-v error -stats, turned off all output except for errors and left a line about encoding
-i pipe:0, at first I thought that each camera should have its own number, but it's not! Every camera should have 0 everywhere, I have three cameras...

David, thanks for the support, now everything works as it should! Sorry again for being such a donkey. Topic can be marked as solved!
